### ISSUE TITLE
Timeout from Pool- possible crosstalk between different pools in same application?

### ISSUE THREAD

**Issue by amanullah92** [brettwooldridge/HikariCP#1236]
#### Environment
```
HikariCP version: 3.1.0
JDK version     : 1.8.0_91
Database        : MySQL
Driver version  : 8.0.11
```

After upgrading both driver from (5.6.xx) to 8.0.11 and HikariCP from 2.3.5 to 3.1.0 we are seeing intermittent timeouts (once or twice a day) from the connection pool. 

Our application talks to 6 different databases, and each has a separate pool. What we are seeing is that once any particular pool has an issue, like a connection is broken/new connection is being added, sometimes it would disrupt our master database pool, causing timeouts for a few hundred milliseconds before returning to normal. 

In some cases, application doesn't do criss-cross of connections- like acquiring one connection from a pool and waiting on the other pool- this doesn't seem to be the case. See logs below:

12 Sep 2018 23:03:46,092 -[HikariPool-3 housekeeper]-[]- DEBUG - HikariPool.logPoolState(411) - HikariPool-3 - Before cleanup stats (total=2, active=0, idle=2, waiting=0)
12 Sep 2018 23:03:46,924 -[HikariPool-3 housekeeper]-[]- DEBUG - HikariPool.logPoolState(411) - HikariPool-3 - After cleanup  stats (total=2, active=0, idle=2, waiting=0)
12 Sep 2018 23:03:46,096 -[HikariPool-1 housekeeper]-[]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Pool stats (total=10, active=0, idle=10, waiting=0)
**12 Sep 2018 23:03:46,949 -[http-nio-8080-exec-9]-[19109153419:65]-  WARN - PoolBase.isConnectionAlive(175) - HikariPool-2 - Failed to validate connection com.mysql.cj.jdbc.ConnectionImpl@68cc0ffe (No operations allowed after connection closed.)
12 Sep 2018 23:03:46,986 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(129) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@68cc0ffe: (connection is dead)
12 Sep 2018 23:03:47,007 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG -** PoolBase.quietlyCloseConnection(138) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@68cc0ffe failed
	at com.zaxxer.hikari.pool.HikariPool.lambda$closeConnection$1(HikariPool.java:441)
12 Sep 2018 23:03:47,056 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool$PoolEntryCreator.call(709) - HikariPool-2 - Added connection com.mysql.cj.jdbc.ConnectionImpl@2d4038b5
12 Sep 2018 23:03:47,057 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool.logPoolState(411) - HikariPool-2 - After adding stats (total=10, active=0, idle=10, waiting=0)
12 Sep 2018 23:03:47,314 -[http-nio-8080-exec-48]-[1085842046:20]-  WARN - PoolBase.isConnectionAlive(175) - HikariPool-2 - Failed to validate connection com.mysql.cj.jdbc.ConnectionImpl@29f105c5 (No operations allowed after connection closed.)
12 Sep 2018 23:03:47,318 -[http-nio-8080-exec-53]-[1085841999:29]-  WARN - PoolBase.isConnectionAlive(175) - HikariPool-2 - Failed to validate connection com.mysql.cj.jdbc.ConnectionImpl@1892aa25 (No operations allowed after connection closed.)
12 Sep 2018 23:03:47,316 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(129) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@29f105c5: (connection is dead)
12 Sep 2018 23:03:47,332 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(138) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@29f105c5 failed
	at com.zaxxer.hikari.pool.HikariPool.lambda$closeConnection$1(HikariPool.java:441)
12 Sep 2018 23:03:47,332 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(129) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@1892aa25: (connection is dead)
12 Sep 2018 23:03:47,332 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(138) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@1892aa25 failed
	at com.zaxxer.hikari.pool.HikariPool.lambda$closeConnection$1(HikariPool.java:441)
12 Sep 2018 23:03:47,357 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool$PoolEntryCreator.call(709) - HikariPool-2 - Added connection com.mysql.cj.jdbc.ConnectionImpl@6c7d328e
12 Sep 2018 23:03:47,380 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool$PoolEntryCreator.call(709) - HikariPool-2 - Added connection com.mysql.cj.jdbc.ConnectionImpl@5ddbdcc5
12 Sep 2018 23:03:47,383 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool.logPoolState(411) - HikariPool-2 - After adding stats (total=10, active=1, idle=9, waiting=0)
12 Sep 2018 23:03:47,439 -[http-nio-8080-exec-3]-[19109179197:7]-  WARN - PoolBase.isConnectionAlive(175) - HikariPool-2 - Failed to validate connection com.mysql.cj.jdbc.ConnectionImpl@296500f0 (No operations allowed after connection closed.)
12 Sep 2018 23:03:47,466 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(129) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@296500f0: (connection is dead)
12 Sep 2018 23:03:47,467 -[HikariPool-2 connection closer]-[19109153419:65]- DEBUG - PoolBase.quietlyCloseConnection(138) - HikariPool-2 - Closing connection com.mysql.cj.jdbc.ConnectionImpl@296500f0 failed
	at com.zaxxer.hikari.pool.HikariPool.lambda$closeConnection$1(HikariPool.java:441)
1**2 Sep 2018 23:03:47,480 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool$PoolEntryCreator.call(709) - HikariPool-2 - Added connection com.mysql.cj.jdbc.ConnectionImpl@c52cdc1
12 Sep 2018 23:03:47,511 -[HikariPool-2 connection adder]-[19109153419:65]- DEBUG - HikariPool.logPoolState(411) - HikariPool-2 - After adding stats (total=10, active=3, idle=7, waiting=0)
12 Sep 2018 23:03:47,787 -[http-nio-8080-exec-92]-[1085842478:16]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=41)
12 Sep 2018 23:03:47,788 -[ThreadPoolTaskExecutor-6]-[19084462804:3]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=41)**
12 Sep 2018 23:03:47,788 -[http-nio-8080-exec-81]-[1680751179:77]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=40)
12 Sep 2018 23:03:47,788 -[http-nio-8080-exec-37]-[1680751016:7]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=40)
12 Sep 2018 23:03:47,789 -[http-nio-8080-exec-30]-[1680753095:21]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=41)
12 Sep 2018 23:03:47,791 -[http-nio-8080-exec-89]-[19109186187:12]- DEBUG - HikariPool.logPoolState(411) - HikariPool-1 - Timeout failure stats (total=10, active=10, idle=0, waiting=41)

-----------------------------------------------------------------------------------------
*Have you searched the CLOSED issues already? How about checking [stackoverflow](https://stackoverflow.com/search?q=HikariCP)?*

**Comment by brettwooldridge** [comment]
Are you tracking pool statistics? Such as average time a connection is out of the pool?  41 waiting threads indicates that queries are running very slow in that pool...

**Comment by brettwooldridge** [comment]
@amanullah92 The message ``Failed to validate connection com.mysql.cj.jdbc.ConnectionImpl@68cc0ffe (No operations allowed after connection closed.)`` typically indicates that the database-side is closing connections before they have reached their ``maxLifetime``.

The fact that the validation failures in *HikariPool-2* at *23:03:46,949* thru *23:03:47,439* are so  near the timeouts experienced in *HikariPool-1* starting at *23:03:47,787* is "suspicious".  In the legal field there is a maxim that *there are no coincidences*.

Having inspected the code, I find no *static* state that could be shared between pools.  My suspicions fall on the database, and why connections were closed (from the database-side) coincidentally close to the timeouts.  Have you checked the database logs at the same timestamp?  (Both servers use NTP, right?!)

It would be useful to take a connection that failed validation, for example ``ConnectionImpl@68cc0ffe``, and look back in the log to find when it was created.  What you are looking for is: what is the time difference between creation and validation failure?  I would suggest trying to find the *shorted* time difference between *creation* and *validation failure*, and set the ``maxLifetime`` to a minute or so shorter than that time.

**Comment by amanullah92** [comment]
Curiously enough, Pool-1 connects to master, while Pool-2 connects to HAProxy which has three slaves behind it. So the pools don't even share the database! 

I'm working on removing the failed validation errors in Pool-2 by reducing the maxLifetime as you suggested, let's see what we get after this.

**Comment by brettwooldridge** [comment]
This almost seems to indicate synchronization within the driver, if the databases are truly separate...

**Comment by brettwooldridge** [comment]
Any update on this issue?

**Comment by amanullah92** [comment]
Yes, we removed the cross talk angle by making sure validation failures don't occur- but the intermittent timeouts remained. 

It turned out that there was an anti-virus (TrendMicro Deep Security Agent) upgrade on the machines which was causing this. Not sure exactly how- we are following up with their team on this. But after disabling the anti-virus, we got rid of this behaviour.

Will update here when we get to the root of this. For now, we are pretty happy to get rid of this!
